### ISSUE TITLE
Don't offer IDE0018 (Variable declaration can be inlined) for use of same out variable multiple times in same scope

### ISSUE THREAD

**Issue by hughbe** [dotnet/roslyn#18344]
### Input
```
DateTimeOffset result;
Assert.True(DateTimeOffset.TryParse("4/21 5am", new CultureInfo("en-US"), DateTimeStyles.None, out result));
Assert.Equal(4, result.Month);
Assert.Equal(21, result.Day);
Assert.Equal(5, result.Hour);

Assert.True(DateTimeOffset.TryParse("4/21 5pm", new CultureInfo("en-US"), DateTimeStyles.None, out result));
Assert.Equal(4, result.Month);
Assert.Equal(21, result.Day);
Assert.Equal(17, result.Hour);
```

### Actual
```
Assert.True(DateTimeOffset.TryParse("4/21 5am", new CultureInfo("en-US"), DateTimeStyles.None, out 
DateTimeOffset  result));
Assert.Equal(4, result.Month);
Assert.Equal(21, result.Day);
Assert.Equal(5, result.Hour);

Assert.True(DateTimeOffset.TryParse("4/21 5pm", new CultureInfo("en-US"), DateTimeStyles.None, out result));
Assert.Equal(4, result.Month);
Assert.Equal(21, result.Day);
Assert.Equal(17, result.Hour);
```

### Expected
IDE refactoring not offered

I guess this is kind of subjective, but it's confusing as it's hard to see where the variable is declared

**Comment by Pilchie** [comment]
@CyrusNajmabadi @kuhlenh - thoughts here?

**Comment by CyrusNajmabadi** [comment]
To me, the point is that we're offering a suggestion. At the end of the day, it's only telling you what you can do.  It's not telling you "you have to do this".  If the suggestion isn't something you want, then i think it's fine to not listen to it.

FOr example, in Roslyn itself i get lots of suggestions to not use "this." when it's not necessary.  but in some places i like it, so i just keep them.

--

Note: i'm thinking that certain 'style' suggestions might be ones that really can't be elevated past the suggestion point.  i.e. saying to someone "it's an error if you don't use a ?? expression or you don't use an out-var" doesn't really make sense to me.  These are not 'style' per-se, but more thoughtful ideas on things you could do if you were so inclined.

**Comment by kuhlenh** [comment]
@Pilchie I think this is where different icons can really come in handy. Rather than differentiating between "refactoring v code fix" maybe we should think of differentiating between "_should_ fix and _can_ fix". That way, developers can know when to investigate feedback and when to ignore it until later.

I'm not quite sure how this works with team conventions in place. You could imagine we would introduce a "hint" severity level which has a different icon than the lightbulb and is used only for "can fix" items. This would allow the yellow lightbulb to represent the "should fix" items at their varying degrees of severity.

**Comment by CyrusNajmabadi** [comment]
I like that idea.  We really only should be conveying "should fix" for warnings/errors.  For just suggestions/cleanup, it's a "can fix" notification.

**Comment by hughbe** [comment]
I think for something like this is should be offered as a refactoring (through the lightbulb), but shouldn't be underlined or visible in the error messages section

**Comment by sharwell** [comment]
I also would like to see improvements surrounding suggestions relating to "should" vs. "can". However, as far as I can tell the IDE is currently behaving as intended for the particular refactoring option provided for IDE0018. I would recommend marking this as **Resolution-By Design** and if someone has an idea for specific was to improve the presentation across a variety of available cleanup and refactoring options I would encourage them to file a new bug for that.

**Comment by CyrusNajmabadi** [comment]
>I think for something like this is should be offered as a refactoring (through the lightbulb), but shouldn't be underlined or visible in the error messages section

@hughbe You can control the visibility of these sorts of things through:

Tools | Options | text Editor | C# | Code Style | Prefer Inlined Variable Declaration

If you don't want underlining or visibility in the error messages section, just set the option to:

![image](https://cloud.githubusercontent.com/assets/4564579/24581279/44bcbc66-16cd-11e7-8c59-2b774e586d05.png)

We think, by default, it's helpful to let people know about these things.  But we definitely want you to be able to turn it off if it's not helpful to you.

**Comment by kindermannhubert** [comment]
This is very undesirable while using "Text Editor/C#/Advanced/Enable full solution analyses". Within our team we use only analyzers which you have to respect. So these analyzers cannot report false positives. And we like this analyzer so turning it off is not an option. We want to have common style throughout whole code base.

There is unpleasant inconsistency at variable availability in the same scope between 'if' statements and other statements (not sure why) and IDE0018 works correctly only with 'if' statements in this sense.

For example if we have following methods M1, M2 and M3:
```C#
static void M1()
{
    int x;
    if (int.TryParse("1", out x)) { }
    int.TryParse("2", out x);
}

static void M2()
{
    int x;
    while (int.TryParse("1", out x)) { }
    int.TryParse("2", out x);
}

static void M3()
{
    int x;
    using (GetDisposable(out x)) { }
    int.TryParse("2", out x);
}

static IDisposable GetDisposable(out int x)
{
    x = 1;
    return null;
}
```
than IDE0018 is detected for all 'int x;' declarations but it is valid only in method M1. Fixing it at methods M2 and M3 leads to uncompilable code.

**Comment by CyrusNajmabadi** [comment]
@kindermannhubert  The issue you're encountering has been fixed with: https://github.com/dotnet/roslyn/pull/18109

Here's the experience you get now:

![image](https://cloud.githubusercontent.com/assets/4564579/24996901/85521134-1fe9-11e7-9e2a-2d6691d2039b.png)

**Comment by CyrusNajmabadi** [comment]
@kindermannhubert 

> This is very undesirable while using "Text Editor/C#/Advanced/Enable full solution analyses". Within our team we use only analyzers which you have to respect. So these analyzers cannot report false positives.

This current issue is not about false-positives.  It's about someone just not liking the suggestion at all, even though it's totally valid.  If you are getting false-positives, then that's something we'll always want to look at (though we've fixed a whole host of them, and you should get those fixes in the next VS update).  If you encounter these, please file *new* issues on them.  

Thanks!

**Comment by kindermannhubert** [comment]
Thanks for explanation. I was talking about false positives primarily because of example above (wrong fix of methods M2 and M3) which is evidently already fixed.

**Comment by CyrusNajmabadi** [comment]
@kindermannhubert Yup.  I totally understand :)   I just don't want tangentially related concerns being cross-discussed in the same issue. Let's leave this issue around the experience for people who do/don't want the feature (even when it's not a false-positive), and move any false positive reports to new issues.

Thanks!

**Comment by AndrewMutation** [comment]
@CyrusNajmabadi

Thanks for the hint about how to find the option for IDE0018.

I've just upgraded to a version of VS with both light-bulb and little screw driver quick actions. 

It seems like there are a lot more of them showing up and it feels like the signal to noise ratio isn't great but I don't want to just ignore them all.

Maybe the IDE experience for configuring them is what needs to be changed.

:)

-Andrew
